File: hub/src/midi/midi_filter.py
import time

class MidiFilter:
    noteHistoryLength = 1.0
    doubleHitThreshold = 0.1

    def __init__(self):
        self.recentNotes = []

    def filterNote(self, note):
        self.cleanHistory()
        note = self.filterOutNoteOff(note)
        note = self.filterOutDoubleHits(note)
        if note:
            self.recentNotes.append(note)

        return note

    def cleanHistory(self):
        t = time.time()
        self.recentNotes = list(filter(lambda n: t - n.time < self.noteHistoryLength, self.recentNotes))

    def filterOutDoubleHits(self, note):
        if not note:
            return None

        t = time.time()
        duplicates = list(filter(lambda n: note.note == n.note and t - n.time < MidiFilter.doubleHitThreshold, self.recentNotes))
        if len(duplicates) > 0:
            return None
        return note

    def filterOutNoteOff(self, note):
        if not note:
            return None

        if not note.isNoteOn():
            return None
        return note

File: hub/drumlite-js/src/midi/MidiFilter.ts
import MidiDrumNote from "./MidiDrumNote";

export default class MidiFilter {
    recentNotes: MidiDrumNote[] = [];
    
    static readonly noteHistoryLengthMillis = 1000;
    static readonly doubleHitThresholdMillis = 100;

    public filterNote(note: MidiDrumNote) {
        this.cleanHistory();
        let filteredNote = this.filterOutNoteOff(note);
        filteredNote = this.filterOutNoteOff(filteredNote);
        filteredNote = this.filterOutDoubleHits(filteredNote);
        if (filteredNote) {
            this.recentNotes.push(filteredNote)
        }

        return filteredNote
    }

    private cleanHistory() {
        const t = new Date();
        this.recentNotes = this.recentNotes.filter(n => t.getTime() - n.time.getTime() < MidiFilter.noteHistoryLengthMillis, this.recentNotes);
    }

    private filterOutDoubleHits(note?: MidiDrumNote): MidiDrumNote | undefined {
        if (!note) {
            return undefined;
        }

        const t = new Date();
        const duplicates = this.recentNotes.filter(n => note.note == n.note && t.getTime() - n.time.getTime() < MidiFilter.doubleHitThresholdMillis);
        if (duplicates.length > 0) {
            return undefined;
        }
        return note;
    }

    private filterOutNoteOff(note?: MidiDrumNote): MidiDrumNote | undefined {
        if (!note) {
            return undefined;
        }

        if (!note.isNoteOn()) {
            return undefined;
        }
        return note;
    }
}